### PR TITLE
GO SDK for DHCP based NSX-T segment

### PR DESCRIPTION
**Pull request by Pushpalatha555**

### CODE REVIEW

File: pkg/client/networks.go [hunk 1]
```diff
@@ -61,10 +61,6 @@ func (n *NetworksAPIService) CreateNetwork(
 	networkReq models.CreateNetworkRequest,
 ) (models.CreateNetworkResponse, error) {
 	var networksResp models.CreateNetworkResponse
-	if v, _ := parseVersion("5.4.4"); v <= n.Client.getVersion() {
-		networkReq.Network.NetworkPool.Pool = networkReq.Network.PoolID
-		networkReq.Network.PoolID = 0
-	}
```
Comment (manjunath-batakurki) [inline pkg/client/networks.go:67]: why did you remove this ?

File: pkg/models/networks.go [hunk 2]
```diff
@@ -77,38 +77,35 @@ type CreateNetwork struct {
...
 	DisplayName         string               `json:"displayName,omitempty" tf:"display_name"`
-	CloudID             int                  `json:"-" tf:"cloud_id"`
-	GroupID             string               `json:"-" tf:"group_id"`
-	TypeID              int                  `json:"-" tf:"type_id"`
-	PoolID              int                  `json:"pool,omitempty" tf:"pool_id"`
-	NetworkDomainID     int                  `json:"-" tf:"domain_id"`
-	Site                IDStringModel        `json:"site,omitempty"`
-	Type                IDModel              `json:"type,omitempty"`
-	NetworkDomain       *IDModel             `json:"networkDomain,omitempty"`
-	NetworkProxy        *IDModel             `json:"networkProxy,omitempty"`
```
Comment (manjunath-batakurki) [inline pkg/models/networks.go:88]: why did you remove these ? We still have support for Static Pool Networks. Did the story say that we need to remove static pool networks anywhere ?

File: pkg/models/networks.go [hunk 3]
```diff
@@ -77,38 +77,35 @@ type CreateNetwork struct {
...
+	NetworkServer       IDModel              `json:"networkServer" tf:"network_server"`
 	SearchDomains       string               `json:"searchDomains,omitempty" tf:"search_domains"`
-	Cidr                string               `json:"cidr,omitempty" tf:"cidr"`
+	Cidr                string               `json:"cidr" tf:"gateway_cidr"`
 	Gateway             string               `json:"gateway,omitempty" tf:"gateway"`
 	DNSPrimary          string               `json:"dnsPrimary,omitempty" tf:"primary_dns"`
 	DNSSecondary        string               `json:"dnsSecondary,omitempty" tf:"secondary_dns"`
-	Config              *CreateNetworkConfig `json:"config,omitempty" tf:"config,sub"`
+	Config              *CreateNetworkConfig `json:"config" tf:"config,sub"`
 	Active              bool                 `json:"active" tf:"active"`
```
Comment (manjunath-batakurki) [inline pkg/models/networks.go:91]: my only worry about changing the tf fields is that it could break the existing automation if any of the teams have used ? what is the need to change the tf fields which were already existing ? Please have a good explanation about this.

File: pkg/models/shared_models.go [hunk 4]
```diff
@@ -5,11 +5,11 @@ package models
 import "encoding/json"
 
 type IDModel struct {
-	ID int `json:"id,omitempty" tf:"id"`
```
Comment (manjunath-batakurki) [inline pkg/models/shared_models.go:8]: this is being used in many places, I hope this doesnt break anything. why do we need to remove omit_empty here ?

File: pkg/models/networks.go (hunk 3 above)
Comment (manjunath-batakurki) [reply pkg/models/networks.go:91]: Use can create both DHCP and Static Networks. Ideally he should be able to create one at a time. Please add necessary validations for the same.
https://turbo-fortnight-4dcdc26d.pages.github.io/#tag/Networks/operation/CreatesANetwork

File: pkg/client/networks.go (hunk 1 above)
Comment (Pushpalatha555) [reply pkg/client/networks.go:67]: updated

File: pkg/models/networks.go (hunk 2 above)
Comment (Pushpalatha555) [reply pkg/models/networks.go:88]: updated

File: pkg/models/networks.go (hunk 3 above)
Comment (Pushpalatha555) [reply pkg/models/networks.go]: updated the proper model for both  DHCP and Static Networks.

File: pkg/models/shared_models.go (hunk 4 above)
Comment (Pushpalatha555) [reply pkg/models/shared_models.go]: updated

File: pkg/models/networks.go [hunk 5]
```diff
@@ -78,37 +78,72 @@ type CreateNetwork struct {
...
 	ResourcePermissions NetworkResPermission `json:"-" tf:"resource_permissions,sub"`
 }
 
+type CreateDhcpNetwork struct {
+	DhcpServer bool           `json:"-" tf:"dhcp_enabled"`
+	Config     *NetworkConfig `json:"-" tf:"config,sub"`
+}
+
+type CreateStaticNetwork struct {
+	ExternalID string         `json:"-" tf:"external_id,computed"`
```
Comment (manjunath-batakurki) [inline pkg/models/networks.go:121]: why do need these to be tf fields ?

Comment (manjunath-batakurki) [reply pkg/models/networks.go:121]: InternalID, Unique ID , External ID

Comment (manjunath-batakurki) [reply pkg/models/networks.go:121]: code as well.

File: pkg/models/networks.go [hunk 6]
```diff
@@ -78,37 +78,72 @@ type CreateNetwork struct {
...
 type CreateNetworkConfig struct {
-	ConnectedGateway string `json:"connectedGateway,omitempty" tf:"connected_gateway"`
-	VlanIDs          string `json:"vlanIDs,omitempty" tf:"vlan_ids"`
+	ConnectedGateway        string `json:"connectedGateway,omitempty"`
+	VlanIDs                 string `json:"vlanIDs,omitempty"`
+	SubnetIPManagementType  string `json:"subnetIpManagementType"`
+	SubnetIPServerID        string `json:"subnetIpServerId"`
+	SubnetDhcpServerAddress string `json:"subnetDhcpServerAddress"`
+	DhcpRange               string `json:"dhcpRange"`
+	SubnetDhcpLeaseTime     string `json:"subnetDhcpLeaseTime"`
```
Comment (manjunath-batakurki) [inline pkg/models/networks.go:136]: why is lease time a string ?

**Review by manjunath-batakurki — COMMENTED** [review]
Minor comments, looks good.

File: pkg/models/networks.go (hunk 6 above)
Comment (Pushpalatha555) [reply pkg/models/networks.go:136]: yes

File: pkg/models/networks.go (hunk 5 above)
Comment (Pushpalatha555) [reply pkg/models/networks.go:121]: before  tf fields was in the line https://github.com/HewlettPackard/hpegl-vmaas-cmp-go-sdk/blob/77c6f87dd327ec7efa15bdaabf62e689d5e5f9a9/pkg/models/networks.go#L35 now I moved to static block

**Review by manjunath-batakurki — APPROVED** [review]

**Review by ravikumar29524 — APPROVED** [review]
